fields.3
.\"
.\" $Id: fields.3,v 1.5 2021-06-02 16:52:45-07 geoff Exp $
.\"
.\" $Log: fields.3,v $
.\" Revision 1.5  2021-06-02 16:52:45-07  geoff
.\" Remove documentation for obsolete features.  Note that maxf and
.\" nfields are unsigned.
.\"
.\" Revision 1.4  2021-05-29 22:58:29-07  geoff
.\" Fix some minor typos.
.\"
.\" Revision 1.3  1994/01/05  20:13:43  geoff
.\" Add the maxf parameter
.\"
.\" Revision 1.2  1994/01/04  02:40:16  geoff
.\" Add descriptions of field_line_inc, field_field_inc, and the
.\" FLD_NOSHRINK flag.
.\"
.\" Revision 1.1  1993/09/09  01:09:44  geoff
.\" Initial revision
.\"
.\" 
.TH FIELDS 3 local
.SH NAME
fieldread, fieldmake, fieldwrite, fieldfree \- field access package
.SH SYNTAX
.nf
#include "fields.h"

typedef struct {
    unsigned int nfields;
    int hadnl;
    char *linebuf;
    char **fields;
} field_t;

#define FLD_RUNS        0x0001
#define FLD_SNGLQUOTES  0x0002
#define FLD_BACKQUOTES  0x0004
#define FLD_DBLQUOTES   0x0008
#define FLD_SHQUOTES    0x0010
#define FLD_STRIPQUOTES 0x0020
#define FLD_BACKSLASH   0x0040

extern field_t *fieldread (FILE * file, char * delims,
                           int flags, unsigned int maxf);
extern field_t *fieldmake (char * line, int allocated,
                           char * delims, int flags, unsigned int maxf);
extern int fieldwrite (FILE * file, field_t * fieldp,
                       int delim);
extern void fieldfree (field_t * fieldp);
.fi
.SH DESCRIPTION
.PP
The fields-access package eases the common task of parsing and
accessing information that is separated into fields by whitespace or
other delimiters.  Various options can be specified to handle many
common cases, including selectable delimiters, runs of delimiters, and
quoting.
.PP
.I fieldread
reads one line from a file, parses it into fields as specified by the
parameters, and returns a
.B field_t
structure describing the result.
.I fieldmake
performs the same process on a buffer already in memory.
.I fieldwrite
creates an output line from a
.B field_t
structure and writes it to an output file.
.I fieldfree
frees a
.B field_t
structure and any associated memory allocated by the package.
.PP
The
.B field_t
structure describes the fields in a parsed line.
A well-behaved program should only access the
.BR nfields ,
.BR fields ,
and
.B hadnl
elements;
all other elements are used internally by the package and are not
guaranteed to remain the same even though they are documented here.
.B Nfields
gives the number of fields in the parsed line, just like the
.B argc
argument to a C program;
.B fields
is a pointer to an array of string pointers, just like the
.B argv
argument to a C program.
As in C, the last field pointer is followed by a null pointer,
although the field count is the preferred method of accessing fields.
The user may alter
.B nfields
by decreasing it, and may replace any pointer in
.B fields
without harm.
This is often useful in replacing a single field with a calculated
value preparatory to output.
The
.B hadnl
element is nonzero if the original line was terminated with a newline
when it was parsed;
this is used to accurately reproduce the input when
.I fieldwrite
is called.
.PP
The
.B linebuf
element contains a pointer to an internal buffer allocated by
.I fieldread
or provided to
.IR fieldmake .
This buffer is
.I not
guaranteed to contain anything sensible, although in the current
implementation all of the field contents can be found therein.
.PP
.I fieldread
reads a single line of arbitrary length from
.BR file ,
allocating as much memory as necessary to hold it, and then parses the
line according to its remaining arguments.
A pointer to the parsed
.B field_t
structure is returned, with
.B NULL
returned if an error occurs or if
.B EOF
is reached on the input file.
Fields in the input line are considered to be separated by any of the
delimiters in the
.B delims
parameter.
For example, if delimiters of ":.;" are specified, a line containing
"a:b;c.d" would be considered to have four fields.
.PP
The default parsing of fields considers each delimiter to indicate a
separate field, and does not allow any quoting.  This is similar to
the parsing done by
.IR cut (1).
This behavior can be modified by specifying flags.
Multiple flags may be OR'ed together.
The available flags are:
.IP \fBFLD_RUNS\fP
Consider runs of delimiters to be the same as a single delimiter,
suppressing all null fields.
This is similar to the way utilities like
.IR awk (1)
and
.IR sort (1)
treat whitespace, but it is not limited to whitespace.
A run does not have to consist of a single type of delimiter;  if both
semicolon and colon are delimiters, ";::;" is a run.
.IP \fBFLD_SNGLQUOTES\fP
Allow field contents to be quoted with single quotes.
Delimiters and other quotes appearing within single quotes are ignored.
This may appear in combination with other quote options.
.IP \fBFLD_BACKQUOTES\fP
Allow field contents to be quoted with reverse single quotes.
Delimiters and other quotes appearing within reverse single quotes are ignored.
This may appear in combination with other quote options.
.IP \fBFLD_DBLQUOTES\fP
Allow field contents to be quoted with single quotes.
Delimiters and other quotes appearing within double quotes are ignored.
This may appear in combination with other quote options.
.IP \fBFLD_SHQUOTES\fP
Allow shell-style quoting.
In the absence of this option, quotes are only recognized at the
beginning of a field, and characters following the close quote are
removed from the field (and are thus lost from the input line).
If this option is specified, quotes may appear within a field, in the
same way as they are handled by
.IR sh (1).
Multiple quoting styles may be used in the same field.
If none of
.BR FLD_SNGLQUOTES ,
.BR FLD_BACKQUOTES ,
or
.B FLD_DBLQUOTES
is specified with
.BR FLD_SHQUOTES ,
all three options are implied.
.IP \fBFLD_STRIPQUOTES\fP
Remove quotes and backslash sequences from the field while parsing,
converting backslash sequences to their proper ASCII equivalent.
The C sequences \ea, \eb, \ef, \en, \er, \ev, \ex\fInn\fP, and \e\fInnn\fP are
supported.
Any other sequence is simply converted to the backslashed character,
as in
.IR sh (1).
.IP \fBFLD_BACKSLASH\fP
Accept standard C-style backslash sequences.
The sequence will be converted to an ASCII equivalent if
.B FLD_STRIPQUOTES
is specified (q.v.).
.PP
The
.I maxf
parameter, if nonzero, specifies the maximum number of fields to be
generated.
This may enhance performance if only the first few fields of a long
line are of interest to the caller.
The actual number of fields returned is one greater than
.IR maxf ,
because the remainder of the line will be returned as a single
contiguous (and uninterpreted, 
even if
.B FLD_STRIPQUOTES
or
.B FLD_BACKSLASH
is specified) field.
.PP
.I fieldmake
operates exactly like
.IR fieldread ,
except that the line parsed is provided by the caller rather than
being read from a file.
If the
.I allocated
parameter is nonzero, the memory pointed to by the
.I line
parameter will automatically be freed when
.I fieldfree
is called;
otherwise this memory is the caller's responsibility.
The memory pointed to by
.I line
is destroyed by
.IR fieldmake .
All other parameters are the same as for
.IR fieldread.
.PP
.I fieldwrite
writes a set of fields to the specified
.IR file ,
separating them with the delimiter character
.I delim
(note that this is a character, not a string), and appending a newline
if specified by the
.I hadnl
element of the structure.
The field structure is not freed.
.I fieldwrite
will return nonzero if an I/O error is detected.
.PP
.I fieldfree
frees the
.B field_t
structure passed to it, along with any associated auxiliary memory
allocated by the package (or passed to
.IR fieldmake ).
The structure may not be accessed after
.I fieldfree
is called.
.SH RETURN VALUES
.I fieldread
and
.I fieldmake
return
.B NULL
if an error occurs or if
.B EOF
is reached on the input file.
.I fieldwrite
returns nonzero if an output error occurs.
.SH BUGS
Thanks to the vagaries of ANSI C, the
.B fields.h
header file defines an auxiliary macro named
.BR P .
If the user needs a similarly-named macro, this macro must be
undefined first, and the user's macro must be defined after
.B fields.h
is included.


CHANGES
Version 3.4.02
==============

 - A major bug in 3.4.01 that caused incorrect behavior with many
   languages has been corrected.

 - The default maximum number of stringchars has been doubled.
 
 - Some spurious compiler warnings have been suppressed.

Version 3.4.01
==============

 - The english.aff file has been revised so that UTF-8 is now the
   default encoding for English dictionaries.

 - Some subtle bugs with the handling of stringchars, which were
   revealed by the changes to english.aff, have been corrected.

 - A new affix command, "options", has been added to allow selection
   of options related to a particular string character type.  There
   are currently two options, raw_display and squeeze_strings.  Both
   should be activated for UTF-8 encodings but few others.  This
   change causes hash files built with earlier ispell versions to be
   incompatible; old hash files will need to be rebuilt.

 - Over 6000 words have been added to the English dictionaries.  Most
   of them are in the english.xlg variant, but some are included in
   smaller dictionaries as well.  The English dictionaries also take
   advantage of the new UTF-8 encoding to support correct spellings of
   words adopted from other languages, such as crêpe.

Version 3.4.00
==============

 - The english.5 manual page has been dropped from the distribution;
   the english.aff file provides superior documentation (and in any
   case, most people will use munchlist to apply affixes).

 - Ispell now supports systems (BSD) that foolish discarded backwards
   compatibility and insist on using termios.

 - A function (getline) has been renamed in correct.c to solve
   compilation problems on some systems.

 - The manual page now correctly identifies the ispell version that it
   describes.

 - It is now possible to insert the ispell version information into
   shell scripts via the Makefile.

 - An error has been corrected in the English affix files that caused
   certain words ending in "th" to be pluralized incorrectly.  The
   dictionaries have been updated to ensure that no incorrect plurals
   have crept in.

 - The personal dictionary is now written in a stable order when there
   are multiple variant capitalizations of a word.

 - The security of temporary files has been improved on systems that
   don't have the "mktemp" command.

 - The deformatters makefile has been changed to be compatible with
   older versions of make.

 - A bug in TeX deformatting has been corrected.  Previously, two
   adjacent math-mode environments introduced with dollar signs, such
   as $a=b$$c=d$, would cause ispell to lose track of whether it was
   in math mode.

 - The Makefile now supports a DESTDIR installation prefix (patch from
   Petter Reinholdtsen)

 - The American and British Makefiles have been simplified to ensure
   that dictionaries are always built with the lastest information.

Version 3.3.02
==============

 - A Makefile bug has been corrected which caused the default hash
   file to be created under the wrong name.  If you installed 3.3.01,
   you should remove the file /usr/local/lib/ispell/english (or
   whatever your default language is, and in the appropriate
   directory) to clean up from the effects of this bug.

 - A bug has been corrected that caused ispell to allocate space for
   zero elements when initializing the expansion tables.

 - Hash files are now installed in reverse order of how they are
   listed in LANGUAGES.  As a result, the default american or british
   hash file is now that first listed in LANGUAGES, rather than the
   last.

 - A misspelling in config.X has been corrected.  This would only have
   affected Windows systems that didn't use the supplied local.h
   files.

 - EXEEXT is now printed by "ispell -vv".

 - The local.h.macos file now correctly leaves USG undefined.

 - There is now a local.h.cygwin file.

Version 3.3.01
==============

 - The default dictionary is now the first dictionary defined in
   LANGUAGES.  It is no longer necessary to define MASTERHASH,
   DEFHASH, and DEFLANG.  However, if you define those variables
   there, your definitions will be respected.

 - Count files are no longer used for dictionaries; this corrects
   problems with hash table overflows for some users who build
   multiple dictionaries.

 - If there is a directory named ~/.ispell_logs, ispell in
   command-line mode will create logs of your spelling corrections
   there.  These logs may be useful in identifying your common
   errors.  Someday, they may also be used for research into better
   methods of spelling correction.

 - A bug has been fixed that caused installation to fail because it
   referred to the nonexistent file "fixispell-a".

 - The pc/local.* files have been updated to choose consistent
   hash-file names.

 - A tiny portability problem in ispell.1X has been corrected.

 - DJGPP updates from Eli Zaretskii <eliz@gnu.org>:

	* deformatters/Makefile (PROGRAMS): Use $EXEEXT.

	* config.X (LINK): Define only if undefined.
	(EXEXT): New variable.

	* Makefile (install-basic, install-dictbuild): Use $$EXEEXT for
	systems where executable files have special extensions.
	(install-deformatters): Run config.sh and pass EXEEXT to sub-Make.
	(config.sh): Add EXEEXT to the list of variables put into
	config.sh.

	* pc/make-dj.bat: Sync with the new files and build commands.
	* pc/makeemx.bat: Ditto.

	* pc/djterm.c (tputs, tgetent): Fix return value according to
	prototypes on proto.h.

	* pc/cfglang.sed (PATH_SEPARATOR): Set to ":", since
	english/Makefile prepends to PATH using the ":" separator.

	* pc/README: Update DJGPP repository URL, versions of utilities,
	and other relevant information.

	* pc/local.djgpp: Update Eli Zaretskii's email address.
	(LINK): Set to "cp -p".
	(MAN4DIR): Delete.
	(MAN45DIR): New; define to point to DJGPP's man5.
	(INSTALL): Define to invoke `ginstall'.
	(OPTIONVAR, LIBRARYVAR): Remove, they are handled by config.X.
	(EXEEXT): New; define to ".exe".

fields.c
#ifndef lint
static char Rcs_Id[] =
    "$Id: fields.c,v 1.13 2021-06-02 16:52:07-07 geoff Exp $";
#endif

/*
 * $Log: fields.c,v $
 * Revision 1.13  2021-06-02 16:52:07-07  geoff
 * When maxf is nonzero, don't truncate the last field.  Get rid of
 * O(N^2) performance when expanding buffers.  Remove FLD_NOSHRINK.
 *
 * Revision 1.12  2021-05-29 22:53:37-07  geoff
 * Correctly handle quoted strings in the last field on a line.
 *
 * Revision 1.11  2005-04-14 07:38:23-07  geoff
 * Make maxf unsigned.
 *
 * Revision 1.10  1999/01/06 20:57:19  geoff
 * Include ispell.h so proto.h will work
 *
 * Revision 1.9  1999/01/05  20:46:38  geoff
 * Get declarations from proto.h
 *
 * Revision 1.8  1998/07/06  04:56:19  geoff
 * Make strlen return an unsigned int
 *
 * Revision 1.7  1994/01/06  05:26:37  geoff
 * Get rid of all references to System V string routines, for portability
 * (sigh).
 *
 * Revision 1.6  1994/01/05  20:13:43  geoff
 * Add the maxf parameter
 *
 * Revision 1.5  1994/01/04  02:40:21  geoff
 * Make the increments settable (field_line_inc and field_field_inc).
 * Add support for the FLD_NOSHRINK flag.
 *
 * Revision 1.4  1993/09/27  17:48:02  geoff
 * Fix some lint complaints and some parenthesization errors.
 *
 * Revision 1.3  1993/09/09  01:11:11  geoff
 * Add a return value to fieldwrite.  Add support for backquotes and for
 * unstripped backslashes.
 *
 * Revision 1.2  1993/08/26  00:02:50  geoff
 * Fix a stupid null-pointer bug
 *
 * Revision 1.1  1993/08/25  21:32:05  geoff
 * Initial revision
 *
 */

#include <stdio.h>
#include "config.h"
#include "fields.h"
#include "ispell.h"
#include "proto.h"

field_t *	fieldread P ((FILE * file, char * delims,
				  int flags, unsigned int maxf));
				/* Read a line with fields from a file */
field_t *	fieldmake P ((char * line, int allocated, char * delims,
				  int flags, unsigned int maxf));
				/* Make a field structure from a line */
static field_t * fieldparse P ((field_t * fieldp, char * line, char * delims,
				  int flags, unsigned int maxf));
				/* Parse the fields in a line */
static int	fieldbackch P ((char * str, char ** out, int strip));
				/* Process backslash sequences */
int		fieldwrite P ((FILE * file, field_t * fieldp, int delim));
				/* Write a line with fields to a file */
void		fieldfree P ((field_t * fieldp));
				/* Free a field returned by fieldread */

#define INIT_FIELDS     8       /* Number of fields to allocate initially */
#define INIT_LINE       128     /* Initial size of a line in fieldread */

#ifndef USG
#define strchr	index
#endif /* USG */

/*
 * Read one line of the given file into a buffer, break it up into
 * fields, and return them to the caller.  The field_t structure
 * returned must eventually be freed with fieldfree.
 */
field_t * fieldread (file, delims, flags, maxf)
    FILE *		file;	/* File to read lines from */
    char *		delims;	/* Characters to use for field delimiters */
    int			flags;	/* Option flags;  see fields.h */
    unsigned int	maxf;	/* Maximum number of fields to parse */
    {
    register char *	linebuf; /* Buffer to hold the line read in */
    int			linemax; /* Maximum line buffer size */
    int			linesize; /* Current line buffer size */

    linebuf = (char *) malloc (INIT_LINE);
    if (linebuf == NULL)
	return NULL;
    linemax = INIT_LINE;
    linesize = 0;
    /*
     * Read in the line.
     */
    while (fgets (&linebuf[linesize], linemax - linesize, file)
      != NULL)
	{
	linesize += strlen (&linebuf[linesize]);
	if (linebuf[linesize - 1] == '\n')
	    break;
	else
	    {
	    linemax *= 2;
	    linebuf = (char *) realloc (linebuf, linemax);
	    if (linebuf == NULL)
		return NULL;
	    }
	}
    if (linesize == 0)
	{
	free (linebuf);
	return NULL;
	}
    return fieldmake (linebuf, 1, delims, flags, maxf);
    }

field_t * fieldmake (line, allocated, delims, flags, maxf)
    char *		line;	/* Line to make into a field structure */
    int			allocated; /* NZ if line allocated with malloc */
    char *		delims;	/* Characters to use for field delimiters */
    int			flags;	/* Option flags;  see fields.h */
    unsigned int	maxf;	/* Maximum number of fields to parse */
    {
    register field_t *	fieldp;	/* Structure describing the fields */
    int			linesize; /* Current line buffer size */

    fieldp = (field_t *) malloc (sizeof (field_t));
    if (fieldp == NULL)
	return NULL;
    fieldp->nfields = 0;
    fieldp->linebuf = allocated ? line : NULL;
    fieldp->fields = NULL;
    fieldp->hadnl = 0;
    linesize = strlen (line);
    if (line[linesize - 1] == '\n')
	{
	line[--linesize] = '\0';
	fieldp->hadnl = 1;
	}
    return fieldparse (fieldp, line, delims, flags, maxf);
    }

static field_t * fieldparse (fieldp, line, delims, flags, maxf)
    register field_t *	fieldp;	/* Field structure to parse into */
    register char *	line;	/* Line to be parsed */
    char *		delims;	/* Characters to use for field delimiters */
    int			flags;	/* Option flags;  see fields.h */
    unsigned int	maxf;	/* Maximum number of fields to parse */
    {
    unsigned int	fieldmax; /* Max size of fields array */
    char *		lineout; /* Where to store xlated char in line */
    char		quote;	/* Quote character in use */

    fieldp->nfields = 0;
    fieldmax =
      (maxf != 0  &&  maxf < INIT_FIELDS) ? maxf + 2 : INIT_FIELDS;
    fieldp->fields = (char **) malloc (fieldmax * sizeof (char *));
    if (fieldp->fields == NULL)
	{
	fieldfree (fieldp);
	return NULL;
	}
    if ((flags
	& (FLD_SHQUOTES | FLD_SNGLQUOTES | FLD_BACKQUOTES | FLD_DBLQUOTES))
      == FLD_SHQUOTES)
	flags |= FLD_SNGLQUOTES | FLD_BACKQUOTES | FLD_DBLQUOTES;
    while (1)
	{
	if (flags & FLD_RUNS)
	    {
	    while (*line != '\0'  &&  strchr (delims, *line) != NULL)
		line++;			/* Skip runs of delimiters */
	    if (*line == '\0')
		break;
	    }
	fieldp->fields[fieldp->nfields] = lineout = line;
	/*
	 * Skip to the next delimiter.  At the end of skipping, "line" will
	 * point to either a delimiter or a null byte.
	 */
	if (flags
	  & (FLD_SHQUOTES | FLD_SNGLQUOTES | FLD_BACKQUOTES
	    | FLD_DBLQUOTES | FLD_BACKSLASH))
	    {
	    while (*line != '\0')
		{
		if (strchr (delims, *line) != NULL)
		    break;
		else if (((flags & FLD_SNGLQUOTES)  &&  *line == '\'')
		  ||  ((flags & FLD_BACKQUOTES)  &&  *line == '`')
		  ||  ((flags & FLD_DBLQUOTES)  &&  *line == '"'))
		    {
		    if ((flags & FLD_SHQUOTES) == 0
		      &&  line != fieldp->fields[fieldp->nfields])
			quote = '\0';
		    else
			quote = *line;
		    }
		else
		    quote = '\0';
		if (quote == '\0')
		    {
		    if (*line == '\\'  &&  (flags & FLD_BACKSLASH))
			{
			line++;
			if (*line == '\0')
			    break;
			line += fieldbackch (line, &lineout,
			  flags & FLD_STRIPQUOTES);
			}
		    else
			*lineout++ = *line++;
		    }
		else
		    {
		    /* Process quoted string */
		    if ((flags & FLD_STRIPQUOTES) == 0)
			*lineout++ = quote;
		    ++line;
		    while (*line != '\0')
			{
			if (*line == quote)
			    {
			    if ((flags & FLD_STRIPQUOTES) == 0)
				*lineout++ = quote;
			    line++;		/* Go on past quote */
			    if ((flags & FLD_SHQUOTES) == 0)
				{
				while (*line != '\0'
				  &&  strchr (delims, *line) == NULL)
				    line++;	/* Skip to delimiter */
				}
			    break;
			    }
			else if (*line == '\\')
			    {
			    if (flags & FLD_BACKSLASH)
				{
				line++;
				if (*line == '\0')
				    break;
				else
				    line += fieldbackch (line, &lineout,
				      flags & FLD_STRIPQUOTES);
				}
			    else
				{
				*lineout++ = '\\';
				if (*++line == '\0')
				    break;
				*lineout++ = *line;
				}
			    }
			else
			    *lineout++ = *line++;
			}
		    }
		}
	    }
	else
	    {
	    while (*line != '\0'  &&  strchr (delims, *line) == NULL)
		line++;			/* Skip to delimiter */
	    lineout = line;
	    }
	fieldp->nfields++;
	if (*line++ == '\0')
	    {
	    *lineout = '\0';
	    break;
	    }
	if (maxf != 0  &&  fieldp->nfields > maxf)
	    break;
	*lineout = '\0';
	if (fieldp->nfields >= fieldmax)
	    {
	    fieldmax *= 2;
	    fieldp->fields =
	      (char **) realloc (fieldp->fields, fieldmax * sizeof (char *));
	    if (fieldp->fields == NULL)
		{
		fieldfree (fieldp);
		return NULL;
		}
	    }
	}
    fieldp->fields[fieldp->nfields] = NULL;
    return fieldp;
    }

static int fieldbackch (str, out, strip)
    register char *	str;		/* First char of backslash sequence */
    register char **	out;		/* Where to store result */
    int			strip;		/* NZ to convert the sequence */
    {
    register int	ch;		/* Character being developed */
    char *		origstr;	/* Original value of str */

    if (!strip)
	{
	*(*out)++ = '\\';
	if (*str != 'x'  &&  *str != 'X'  &&  (*str < '0'  ||  *str > '7'))
	    {
	    *(*out)++ = *str;
	    return *str != '\0';
	    }
	}
    switch (*str)
	{
	case '\0':
	    *(*out)++ = '\0';
	    return 0;
	case 'a':
	    *(*out)++ = '\007';
	    return 1;
	case 'b':
	    *(*out)++ = '\b';
	    return 1;
	case 'f':
	    *(*out)++ = '\f';
	    return 1;
	case 'n':
	    *(*out)++ = '\n';
	    return 1;
	case 'r':
	    *(*out)++ = '\r';
	    return 1;
	case 'v':
	    *(*out)++ = '\v';
	    return 1;
	case 'X':
	case 'x':
	    /* Hexadecimal sequence */
	    origstr = str++;
	    ch = 0;
	    if (*str >= '0'  &&  *str <= '9')
		ch = *str++ - '0';
	    else if (*str >= 'a'  &&  *str <= 'f')
		ch = *str++ - 'a' + 0xa;
	    else if (*str >= 'A'  &&  *str <= 'F')
		ch = *str++ - 'A' + 0xa;
	    if (*str >= '0'  &&  *str <= '9')
		ch = (ch << 4) | (*str++ - '0');
	    else if (*str >= 'a'  &&  *str <= 'f')
		ch = (ch << 4) | (*str++ - 'a' + 0xa);
	    else if (*str >= 'A'  &&  *str <= 'F')
		ch = (ch << 4) | (*str++ - 'A' + 0xa);
	    break;
	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	    /* Octal sequence */
	    origstr = str;
	    ch = *str++ - '0';
	    if (*str >= '0'  &&  *str <= '7')
		ch = (ch << 3) | (*str++ - '0');
	    if (*str >= '0'  &&  *str <= '7')
		ch = (ch << 3) | (*str++ - '0');
	    break;
	default:
	    *(*out)++ = *str;
	    return 1;
	}
    if (strip)
	{
	*(*out)++ = ch;
	return str - origstr;
	}
    else
	{
	for (ch = 0;  origstr < str;  ch++)
	    *(*out)++ = *origstr++;
	return ch;
	}
    }

int fieldwrite (file, fieldp, delim)
    FILE *		file;	/* File to write to */
    register field_t *	fieldp;	/* Field structure to write */
    int			delim;	/* Delimiter to place between fields */
    {
    int			error;	/* NZ if an error occurs */
    register unsigned int
			fieldno; /* Number of field being written */

    error = 0;
    for (fieldno = 0;  fieldno < fieldp->nfields;  fieldno++)
	{
	if (fieldno != 0)
	    error |= putc (delim, file) == EOF;
	error |= fputs (fieldp->fields[fieldno], file) == EOF;
	}
    if (fieldp->hadnl)
	error |= putc ('\n', file) == EOF;
    return error;
    }

void fieldfree (fieldp)
    register field_t *	fieldp;	/* Field structure to free */
    {

    if (fieldp == NULL)
	return;
    if (fieldp->linebuf != NULL)
	free ((char *) fieldp->linebuf);
    if (fieldp->fields != NULL)
	free ((char *) fieldp->fields);
    free ((char *) fieldp);
    }
